example/basic/source/app.d

```diff
@@ -16,7 +16,7 @@ struct ExampleBackend {
     enum languageInfo = LanguageInfo("foo", "0.0.1", ".foo");
     int value;
 
-    ExecutionResult execute(in string code) @safe {
+    ExecutionResult execute(in string code, scope IoPubMessageSender sender) @safe {
```

Comment:
Why is a sender parameter needed?

Reply:
When a widget is constructed it needs to send multiple comm_open messages to the frontend. When its state is changed it needs to send a comm_msg, and when its display() is called it needs to send a display_data message.

Imagine the code being:

```
w = widgets.FloatSlider()
w.state.value = 50.0
w.display()
```

Here it needs to send 5 messages. (3x comm_open for FloatSliderModel/LayoutModel/FloatStyleModel, 1 for the state change, and 1 to display).

The point is that there can be an arbitrary number of messages. This doesn't fit the current request/response scheme. Adding a sender sink was the least intrusive I could come up with.

Reply:
A more intrusive but better design is probably needed. At the end of the day the widgets need to send messages, some of those are in response to a request, but that doesn't have to be (e.g. they could have an async datasource, or have an internal timer.)

Reply:
I was thinking about wrapping the Sockets in a refcount and exposing that to the backend. Then it can send whatever it needs to.

Reply:
That makes sense.

Reply:
I propose the following:

- we wrap the sockets in a refcount.
- instead of passing a raw sender delegate we put that in a struct (more robust against changes)
- we keep track of the current message that is being executed
- so that any message send directly via the sockets automatically get the right parentHeader

Reply:
That seems more complicated. I think we'd better move on like you did, then I wait for the next patch when I can see where it's going a bit better and take it from there.